### PR TITLE
PP-1081: Arrayjob 'E' accounting record has the 'start' value set to 0

### PR DESCRIPTION
**Pull request by agrawalravi90**
#### Issue-ID

* **[PP-1081](https://pbspro.atlassian.net/browse/PP-1081)**

#### Problem description
* The E record of an arrayjob would have a correct 'etime' value, but the 'start' value would be set to 0:

11/15/2017 01:23:58;E;4[].pbspro;user=ravi group=wheel project=_pbs_project_default jobname=STDIN queue=workq ctime=1510709028 qtime=1510709028 etime=1510709028 *start=0* array_indices=1-2 Resource_List.ncpus=1 Resource_List.nodect=1 Resource_List.place=pack Resource_List.select=1:ncpus=1 session=0 *end=1510709038* Exit_status=0 run_count=0

#### Cause / Analysis
* The 'stime'/'start' value of an arrayjob was just not being set in the code.

#### Solution description
* Added code to set the stime/start value of an arrayjob when it goes to the "B"(begun) state

#### Checklist:

- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [x] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
- [ ] I have added  **manual test(s) to the Jira ticket and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to the Jira ticket** as evidence of testing/verification.

__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

### CODE REVIEW

File: test/tests/functional/pbs_job_array.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+# organizations to create proprietary derivative works of PBS Pro and
+# distribute them - whether embedded or bundled with other software - under
+# a commercial license agreement.
+#
+# Use of Altair’s trademarks, including but not limited to "PBS™",
+# "PBS Professional®", and "PBS Pro™" and Altair’s logos is subject to Altair's
+# trademark licensing policies.
+
+import calendar
+import time
```
Comment (bhroam) [inline test/tests/functional/pbs_job_array.py]: Time is imported by PTL

File: test/tests/functional/pbs_job_array.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        Check that an arrayjob's E record's 'start' value is set correctly
+        """
+        j = Job(TEST_USER, attrs={
+            ATTR_J: '1-2',
+            'Resource_List.select': 'ncpus=1'
+        })
+        j.set_sleep_time(1)
+
+        # Save the time right now
+        time_now = calendar.timegm(time.gmtime())
```
Comment (bhroam) [inline test/tests/functional/pbs_job_array.py]: Isn't this the same thing as saying int(time.time())?

File: test/tests/functional/pbs_job_array.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        # Extract the 'start' value from the E record
+        values = record.split(";", 3)[3]
+        start_str = " start="
+        values_temp = values.split(start_str, 1)[1]
+        start_val = int(values_temp.split()[0])
+
+        # Check that the difference between 'time_now' and the 'start' val
+        # is less than 5 minutes
+        diff = start_val - time_now
+        self.assertTrue(diff < 300)
```
Comment (bhroam) [inline test/tests/functional/pbs_job_array.py]: Use assertLessThan()

**Comment by agrawalravi90** [comment]
I've addressed your comments Bhroam, please let me know if the code looks fine now. Thanks!

**Comment by agrawalravi90** [comment]
Thanks Bhroam! @all, this a very small PR, 3 line code change with a PTL test, requesting a 2nd reviewer, thanks!

File: test/tests/functional/pbs_job_array.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        values = record.split(";", 3)[3]
+        start_str = " start="
+        values_temp = values.split(start_str, 1)[1]
+        start_val = int(values_temp.split()[0])
+
+        # Check that the difference between 'time_now' and the 'start' val
+        # is less than 60 seconds
+        diff = start_val - time_now
+        self.assertLess(diff, 60, "E record value of 'start' for the " +
+                                  "arrayjob is incorrect")
```
Comment (anamikau) [inline test/tests/functional/pbs_job_array.py]: 60s seems like a huge range. Also 0 or any negative value will also come under this range. why not checking a range something >0 and <20?

Also just wondering this test is more of accounting records, would we rather add to the existing files like 'pbs_acct_log_trunc.py'

Comment (agrawalravi90) [reply test/tests/functional/pbs_job_array.py]: hmm .. I guess I could reduce it, the thing is, I'm calling 'accounting_match' before this with 10 attempts, 1 sec interval each, so, the critical path is at least 10 seconds. How about 30 seconds?

About putting it in a test suite related to accounting logs, I'm kind of torn. This bug didn't exist in the code that directly deals with accounting logs, so I wouldn't call it a test for verifying accounting logs functionality. The bug was inside a piece of code which deals with how arrayjobs function. So, I feel like it is more a test of whether arrayjobs function correctly than whether accounting logs work correctly. Does that make sense?

Comment (anamikau) [reply test/tests/functional/pbs_job_array.py]: okay works for me.

**Comment by agrawalravi90** [comment]
Thanks for the review @anamikau , I changed the time difference to check for to 30 seconds from 60, please let me know if things look okay now.

File: test/tests/functional/pbs_job_array.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+        # Extract the 'start' value from the E record
+        values = record.split(";", 3)[3]
+        start_str = " start="
+        values_temp = values.split(start_str, 1)[1]
+        start_val = int(values_temp.split()[0])
+
+        # Check that the difference between 'time_now' and the 'start' val
+        # is less than 30 seconds
+        diff = start_val - time_now
```
Comment (anamikau) [inline test/tests/functional/pbs_job_array.py]: @agrawalravi90 as discussed please add an explicit check for start time not to be 0. if start time is 0 then above expression will return negative value.

**Comment by agrawalravi90** [comment]
Thanks for catching that Anamika, I've updated the code and attached fresh logs to the ticket, please let me know if it looks okay now.
@bhroam requesting you to take a look at the new changes as well. Thanks!

**Comment by agrawalravi90** [comment]
Thanks anamika!

**Comment by bhroam** [comment]
Looks good @agrawalravi90.  I sign off.  Go ahead and squash and rebase for the merge.

**Comment by agrawalravi90** [comment]
Just did, thanks Bhroam, I'll let you know when the CI is done.

**Comment by bhroam** [comment]
I merged in the PR.

**Pull request merged**